### ISSUE TITLE
Eliminate compiler warnings

### ISSUE THREAD

**Issue by ianhinder** [ScottishCovidResponse/SCRCIssueTracking#152]
Many of the current compiler warnings (when compiling with -Wall) indicate potential bugs.  Some of them are more cosmetic, but if we have many warnings, we might miss important new ones during development.  So I would advocate eliminating all of them, if practical, for all the compilers we are using.

**Comment by WPettersson** [comment]
I've pushed my WIP to dev/compiler-warnings. So far I've only eliminated the switch-default warnings, but there's a rather large caveat. I've added errors (calls to emsg which call exit) to all the default switch cases where doing so does not break bici. However, there are a number of places where a default of "do nothing" might be correct. For instance, in [equation.h line 595](https://github.com/ScottishCovidResponse/CoronaBICI/blob/dev/compiler-warnings/Analysis/header/equation.h#L595), sometimes num1 will not be set. This makes sense, as if the equation is a FUNC (e.g., COSFUNC), then num1 is not used. I've found 20 places where I had to let bici do nothing in the default statement for the code to run.

This leads to two further questions. One, should someone go through and make sure that each of these "empty" default statements is actually correct? And two, should I add code to check for invalid equation types, for instance if ADD is called on only one number? I could just give num1 a default value, but I fear that could just give badly wrong information, and a proper error and exit would serve us better.

**Comment by ianhinder** [comment]
I think that the more internal consistency checks we can add, the better.  Even if they slow the code down a bit, correctness is more important than performance for this project, I would say.

I'm going to comment just on the example you gave, so we have something concrete to discuss.  The main problem here is that there is a possible use of an uninitialised variable (num1) that the compiler cannot prove never happens.  My initial reaction was that we should initialise the variables when they are declared to avoid the undefined behaviour.  However, on reflection, I now think that maybe leaving them uninitialised is good, as long as we trust the compiler to warn us about things like this, and we then give it enough information to know that the uninitialised value is never used.  This avoids the problem that there is no "none" value for doubles that we could safely use.  (I'm not proposing it, but one option would be to use a special "poison" value that would be very visible, e.g. 1.2345e200, which is virtually guaranteed not to arise in practice; I've come across this technique in other codes.  You can then "check for poison" before using a value.)  

Maybe we can wrap the first switch statement in (pseudocode)
```
if (eqncalcop[eq][i] in {ADD, TAKE, MULTIPLY, DIVIDE, POWER}) {
...
}
```
and add the fatal error to the default clause.  That way, num1 is always initialised by the point it is used, and the compiler has enough information to prove that.  If it turns out that eqncalc1[eq][i] has a value not in the cases checked, and num1 is going to be used because it's a binary operation, then this would really have been undefined behaviour, and we will trigger the fatal error.  

Another possible issue: By the fact that you didn't add the fatal error in the default clause of the first switch statement, I conclude that it was being triggered.  That means that for the case of (presumably) binary operations, eqncalc1[eq][i] had a value that was not in the list of cases checked.  I wonder whether this means that it just had a different value, or whether it was actually uninitialised.  A quick glance through the code suggests that it might be initialised with a -1 value (calc1.push_back(-1); calcnum1.push_back(-1)).  Maybe something to check; it's really not good to have things which are really uninitialised floating around.  Though these are vectors of longs, statically allocated, so I think they will be initialised to 0 anyway.

What do you think?

**Comment by WPettersson** [comment]
Yes, you are right that in that particular case, the default clause was being triggered, same holds for anywhere else where the default clause is ` ; // empty`.

My personal preference moving forwards is to re-factor so that values are definitely initialised before use. That would follow your pseudocode idea, something along the lines of

    if (eqncalcop[eq][i] in {ADD, TAKE, MULTIPLY, DIVIDE, POWER}) {
        // perform binary operation
    } else {
        // perform unary operation
    }

with appropriate checks for either one, or both, parameters in the relevant block. To me, that's clearer to read and more obvious. I'll start moving forwards with that now, unless I hear anything to the contrary soon.

**Comment by ianhinder** [comment]
Sounds good!

**Comment by WPettersson** [comment]
One more question: what are your thoughts on turning the function `emsg` (see Analysis/header/var.h) into actual exceptions? It seems both clang and g++ aren't great at noticing that emsg exits the program, and so there are lots of unnecessary "may be used unitialized" complaints that can probably be cleaned up just by using exceptions directly.

One caveat is that I don't know how well this will play with MPI, but at first glance I don't think it will cause problems.

**Comment by ianhinder** [comment]
Would adding [[[noreturn]]](https://en.cppreference.com/w/cpp/language/attributes/noreturn) to the declaration help? I don't think it's a problem to require C++11.

**Comment by WPettersson** [comment]
Would be a smaller code change, for sure.

Also, latest commit on dev/compiler-warnings now has no invalid memory reads under valgrind. There are some small leaks to be plugged, but might be worth testing the MacOS build again to see if it crashes on this branch.

**Comment by ianhinder** [comment]
Excellent!  So you found some actual invalid reads?

**Comment by ianhinder** [comment]
21 iterations of running the code on macOS with your latest branch, and no crashes so far.  That's much better than master, where it would always crash before 10 runs.

**Comment by ianhinder** [comment]
661 iterations overnight with no crashes.  Looks like it's fixed!  I'll take a look at your branch.  Is it ready to merge in?

**Comment by WPettersson** [comment]
You can merge whenever, I'll keep cleaning up these warnings as I go. There are quite a few warnings left.

**Comment by ianhinder** [comment]
I've just compared the results of running for 1000 samples (without MPI) between the current master branch (bd74eee) and the dev/compiler-warnings branch (1350774).  The results are identical apart from a single line in bici0.txt.  The lines in this file are extremely long, but I already see differences at the beginning:
```
394c394
< h|0|369670|369670|369670|0.305826|369670|369670|369670|0.611651|369670|369670|369670|0.917477|369670|369670|369670|1.2233|369670|369670|369670|1.52913|369670|369670|369670|1.83495|369670|369670|369670|2.14078|369670|369670|369670|2.4466|369670|369670|369670|2.75243|369670|369670|369670|3.05826|369670|369670|369670|3.36408|369670|369670|369670|3.66991|369670|369670|369670|3.97573|369670|369670|369670|4.28$
---
> h|0|-|369670|-|0.305826|-|369670|-|0.611651|-|369670|-|0.917477|-|369670|-|1.2233|-|369670|-|1.52913|-|369670|-|1.83495|-|369670|-|2.14078|-|369670|-|2.4466|-|369670|-|2.75243|-|369670|-|3.05826|-|369670|-|3.36408|-|369670|-|3.66991|-|369670|-|3.97573|-|369670|-|4.28156|-|369670|-|4.58738|-|369670|-|4.89321|-|369670|-|5.19903|-|369670|-|5.50486|-|369670|-|5.81069|-|369670|-|6.11651|-|369670|-|6.42234|-|3$
```
There are additional "|-"s between the elements.  Any ideas?

**Comment by ianhinder** [comment]
git bisect says that the change was introduced in
```
1350774f8aabdcb377873b03436e354f47e88ec8 is the first bad commit
commit 1350774f8aabdcb377873b03436e354f47e88ec8
Author: William Pettersson <william@ewpettersson.se>
Date:   Sun May 10 23:10:04 2020 +0100

    If the list is too small, don't get confidence intervals
    
    The list has to have at least 2 entries for the calculation to even
    work, but it's still odd to do CI on such short lists.
```

**Comment by WPettersson** [comment]
Oh, right, yes! Sorry, forgot about those. That's coming from 1350774f8aabdcb377873b03436e354f47e88ec8, and the '|-' are output in cases where a confidence interval cannot be ascertained. Those were just placeholders, I was going to work out what to do when we try to work out confidence intervals on such short lists. I can easily swap it back to do what the old code did, though.

**Comment by ianhinder** [comment]
OK, let's do that, and can you create an issue to track the short list problem?

**Comment by WPettersson** [comment]
I've pushed a commit to just go back to the old code to dev/compiler-warnings, and I'll make that issue now.

**Comment by WPettersson** [comment]
Actually, on the merging comment, do you want me to do some squashing and cleaning up? Or do we not care too much about the merge history at this point?

**Comment by ianhinder** [comment]
I wouldn't worry too much about it.  It's clear from the commits what's going on, I think.
